cdap-spark-core-base/src/main/scala/co/cask/cdap/spark/sql/datasources/stream/StreamRelation.scala

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spark.sql.datasources.stream
+
+import co.cask.cdap.api.common.Bytes
+import co.cask.cdap.api.data.format.{FormatSpecification, StructuredRecord}
+import co.cask.cdap.api.flow.flowlet.StreamEvent
+import co.cask.cdap.api.spark.sql.DataFrames
+import co.cask.cdap.api.stream.GenericStreamEventData
+import co.cask.cdap.app.runtime.spark.SparkClassLoader
+import co.cask.cdap.format.RecordFormats
+import co.cask.cdap.proto.id.StreamId
+import com.google.common.annotations.VisibleForTesting
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SQLContext}
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.util.Try
+
+/**
+  * A BaseRelation and PrunedFilteredScan for reading from Stream as DataFrame.
+  *
+  * By default the stream schema is based on the stream [[co.cask.cdap.api.data.format.FormatSpecification]].
+  * It can be overridden by the `stream.format` parameter with an optional schema provided through the DataFrame API
+  * with all parameters provided to this class will be pass to the [[co.cask.cdap.api.data.format.FormatSpecification]]
+  * as well.
+  * If the `stream.format` is set to `raw`, then the stream body will always be returned as binary data type.
+  *
+  * The timestamp and headers columns will be added automatically to the schema, unless turned off by setting
+  * `timestamp.column.enabled` or `headers.column.enabled` to `false`. The name of the timestamp and headers column
+  * are defaulted to `timestamp` and `headers` respectively, but can be overridden by the parameters
```

Comment:
I also noticed that, although I think `timestamp` is more clear about what it is than `ts`. BTW, it might better be calling it `eventTime` and have the type as SQL timestamp type.